File: server/FONdrum/FONdrum.Domain/Shared/Paginations/Paged.cs
namespace FONdrum.Domain.Shared.Paginations
{
    public record Paged<TEntity>
    {
        public IList<TEntity> Data { get; init; }
        public PageInfo PageInfo { get; init; }

        public Paged(IList<TEntity> data, PageInfo pageInfo)
        {
            if (data == null)
            {
                throw new ArgumentException();
            }

            Data = data;
            PageInfo = pageInfo;
        }

        public Paged(IList<TEntity> data, long totalCount, int? pageSize, int? pageNumber) 
            : this(data, new(totalCount, pageSize, pageNumber))
        {
        }

        public static Paged<TEntity> Of(IList<TEntity> data, PageInfo pageInfo)
            => new(data, pageInfo);

        public static Paged<TEntity> Empty(int? pageSize = null, int? pageNumber = null) 
            => new(new List<TEntity>(), PageInfo.Empty(pageSize, pageNumber));
    }
}

File: web-client/fondrum-web-app/src/app/models/shared/pagination/Paged.ts
import { PageInfo, _defaultPageInfo } from "./PageInfo";

export type Paged<T> = {
  data: T[];
  pageInfo: PageInfo;
};

export function _getDefaultPaged<T>(): Paged<T> {
  return {
    data: [] as T[],
    pageInfo: _defaultPageInfo,
  };
}
